### PR TITLE
gitlab: Add an internal rate limiter

### PR DESCRIPTION
**Pull request by ryanslade**
This adds our own configurable internal rate limiter that is checked before we make API requests to GitLab

Part of: https://github.com/sourcegraph/sourcegraph/issues/8546

### CODE REVIEW

File: internal/extsvc/gitlab/client.go [hunk 1]
```diff
@@ -227,6 +257,15 @@ func (c *Client) do(ctx context.Context, req *http.Request, result interface{})
 		span.Finish()
 	}()
 
+	// Check if we are allowed to proceed based on our rate limiter
+	res := c.RateLimiter.Reserve()
+	if !res.OK() {
+		span.SetTag("internalRateLimit", "exceeded")
+		res.Cancel()
+		return nil, errors.New("internal rate limit exceeded")
```
Comment (unknwon) [inline internal/extsvc/gitlab/client.go:265]: Should we make this an error type? I can imagine we want to special case this error and back off on the background worker.

File: schema/gitlab.schema.json [hunk 2]
```diff
@@ -24,7 +24,29 @@
...
+      "title": "GitLabRateLimit",
+      "required": ["enabled", "requestsPerHour"],
+      "properties": {
+        "enabled": {
+          "description": "true if rate limiting is enabled.",
+          "type": "boolean",
+          "default": true
+        },
+        "requestsPerHour": {
+          "description": "Requests per hour permitted. This is an average, calculated per second.",
```
Comment (unknwon) [inline schema/gitlab.schema.json:39]: https://github.com/sourcegraph/sourcegraph/pull/9606#discussion_r404837748

File: internal/extsvc/gitlab/client.go [hunk 3]
```diff
@@ -235,14 +274,51 @@ func (c *Client) do(ctx context.Context, req *http.Request, result interface{})
...
+	c.rateLimitMonitor.Update(resp.Header)
 	if resp.StatusCode != http.StatusOK {
 		return nil, errors.Wrap(httpError(resp.StatusCode), fmt.Sprintf("unexpected response from GitLab API (%s)", req.URL))
 	}
 
 	return resp.Header, json.NewDecoder(resp.Body).Decode(result)
 }
 
+const (
+	defaultRateLimit      = rate.Limit(10) // 36,000 per hour
```
Comment (unknwon) [inline internal/extsvc/gitlab/client.go:286]: nit: Worthing adding a comment says this is what GitLab allows, so people won't think it's randomly picked.

File: internal/extsvc/gitlab/client.go [hunk 4]
```diff
@@ -235,14 +274,51 @@ func (c *Client) do(ctx context.Context, req *http.Request, result interface{})
...
+)
+
+// Rate limiter cache so that we reuse the rate limiters per
+// external service, even between config changes.
+var limiterMu sync.Mutex
+var limiterCache = make(map[uint32]*rate.Limiter)
+
+func getLimiter(url string, c *schema.GitLabConnection) *rate.Limiter {
+	// Calculate cache key
+	key := fnv1.HashString32(url)
```
Comment (unknwon) [inline internal/extsvc/gitlab/client.go:297]: Why do we only use `url` here but including username and token for bitbucket server? https://github.com/sourcegraph/sourcegraph/pull/9606/files#r404835128

File: internal/extsvc/gitlab/client.go [hunk 5]
```diff
@@ -235,14 +274,51 @@ func (c *Client) do(ctx context.Context, req *http.Request, result interface{})
...
+	defaultRateLimit      = rate.Limit(10) // 36,000 per hour
+	defaultRateLimitBurst = 500
+)
+
+// Rate limiter cache so that we reuse the rate limiters per
+// external service, even between config changes.
+var limiterMu sync.Mutex
+var limiterCache = make(map[uint32]*rate.Limiter)
+
+func getLimiter(url string, c *schema.GitLabConnection) *rate.Limiter {
```
Comment (unknwon) [inline internal/extsvc/gitlab/client.go:295]: The body of this function looks pretty much same to the one we have for bitbucket server, should we unify it?

**Comment by ryanslade** [comment]
I'm going to be trying a different approach to this which I think will simplify things in general

**Pull request closed without merging**